File: src/api/java/ASTVector.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    ASTVector.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/

package com.microsoft.z3;

/**
 * Vectors of ASTs.
 **/
public class ASTVector extends Z3Object {
    /**
     * The size of the vector
     **/
    public int size()
    {
        return Native.astVectorSize(getContext().nCtx(), getNativeObject());
    }

    /**
     * Retrieves the i-th object in the vector.
     * Remarks: May throw an {@code IndexOutOfBoundsException} when 
     * {@code i} is out of range. 
     * @param i Index
     * 
     * @return An AST
     * @throws Z3Exception
     **/
    public AST get(int i)
    {
        return new AST(getContext(), Native.astVectorGet(getContext().nCtx(),
                getNativeObject(), i));
    }

    public void set(int i, AST value)
    {

        Native.astVectorSet(getContext().nCtx(), getNativeObject(), i,
                value.getNativeObject());
    }

    /**
     * Resize the vector to {@code newSize}. 
     * @param newSize The new size of the vector.
     **/
    public void resize(int newSize)
    {
        Native.astVectorResize(getContext().nCtx(), getNativeObject(), newSize);
    }

    /**
     * Add the AST {@code a} to the back of the vector. The size is
     * increased by 1. 
     * @param a An AST
     **/
    public void push(AST a)
    {
        Native.astVectorPush(getContext().nCtx(), getNativeObject(), a.getNativeObject());
    }

    /**
     * Translates all ASTs in the vector to {@code ctx}. 
     * @param ctx A context
     * 
     * @return A new ASTVector
     * @throws Z3Exception
     **/
    public ASTVector translate(Context ctx)
    {
        return new ASTVector(getContext(), Native.astVectorTranslate(getContext()
                .nCtx(), getNativeObject(), ctx.nCtx()));
    }

    /**
     * Retrieves a string representation of the vector.
     **/
    @Override
    public String toString() {
        return Native.astVectorToString(getContext().nCtx(), getNativeObject());
    }

    ASTVector(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    ASTVector(Context ctx)
    {
        super(ctx, Native.mkAstVector(ctx.nCtx()));
    }

    @Override
    void incRef() {
        Native.astVectorIncRef(getContext().nCtx(), getNativeObject());
    }

    @Override
    void addToReferenceQueue() {
        getContext().getASTVectorDRQ().storeReference(getContext(), this);
    }

    /**
     * Translates the AST vector into an AST[]
     * */
    public AST[] ToArray()
    {
        int n = size();
        AST[] res = new AST[n];
        for (int i = 0; i < n; i++)
            res[i] = AST.create(getContext(), get(i).getNativeObject());
        return res;
    }
    
    /**
     * Translates the AST vector into an Expr[]
     * */
    public Expr[] ToExprArray() {
        int n = size();
        Expr[] res = new Expr[n];
        for (int i = 0; i < n; i++)
            res[i] = Expr.create(getContext(), get(i).getNativeObject());
        return res;    
    }

    /**
     * Translates the AST vector into an BoolExpr[]
     * */  
    public BoolExpr[] ToBoolExprArray()
    {
        int n = size();
        BoolExpr[] res = new BoolExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (BoolExpr) Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an BitVecExpr[]
     * */    
    public BitVecExpr[] ToBitVecExprArray()
    {
        int n = size();
        BitVecExpr[] res = new BitVecExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (BitVecExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an ArithExpr[]
     * */   
    public ArithExpr[] ToArithExprExprArray()
    {
        int n = size();
        ArithExpr[] res = new ArithExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (ArithExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an ArrayExpr[]
     * */  
    public ArrayExpr[] ToArrayExprArray()
    {
        int n = size();
        ArrayExpr[] res = new ArrayExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (ArrayExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an DatatypeExpr[]
     * */ 
    public DatatypeExpr[] ToDatatypeExprArray()
    {
        int n = size();
        DatatypeExpr[] res = new DatatypeExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (DatatypeExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an FPExpr[]
     * */   
    public FPExpr[] ToFPExprArray()
    {
        int n = size();
        FPExpr[] res = new FPExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (FPExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an FPRMExpr[]
     * */
    public FPRMExpr[] ToFPRMExprArray()
    {
        int n = size();
        FPRMExpr[] res = new FPRMExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (FPRMExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an IntExpr[]
     * */ 
    public IntExpr[] ToIntExprArray()
    {
        int n = size();
        IntExpr[] res = new IntExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (IntExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }

    /**
     * Translates the AST vector into an RealExpr[]
     * */   
    public RealExpr[] ToRealExprArray()
    {
        int n = size();
        RealExpr[] res = new RealExpr[n];
        for (int i = 0; i < n; i++)
            res[i] = (RealExpr)Expr.create(getContext(), get(i).getNativeObject());
        return res;
    }
}

File: src/api/dotnet/ASTVector.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    ASTVector.cs

Abstract:

    Z3 Managed API: AST Vectors

Author:

    Christoph Wintersteiger (cwinter) 2012-03-21

Notes:
    
--*/

using System;
using System.Diagnostics.Contracts;

namespace Microsoft.Z3
{
    /// <summary>
    /// Vectors of ASTs.
    /// </summary>
    public class ASTVector : Z3Object
    {
        /// <summary>
        /// The size of the vector
        /// </summary>
        public uint Size
        {
            get { return Native.Z3_ast_vector_size(Context.nCtx, NativeObject); }
        }

        /// <summary>
        /// Retrieves the i-th object in the vector.
        /// </summary>
        /// <remarks>May throw an IndexOutOfBoundsException when <paramref name="i"/> is out of range.</remarks>
        /// <param name="i">Index</param>
        /// <returns>An AST</returns>
        public AST this[uint i]
        {
            get
            {
                Contract.Ensures(Contract.Result<AST>() != null);

                return new AST(Context, Native.Z3_ast_vector_get(Context.nCtx, NativeObject, i));
            }
            set
            {
                Contract.Requires(value != null);

                Native.Z3_ast_vector_set(Context.nCtx, NativeObject, i, value.NativeObject);
            }
        }

        /// <summary>
        /// Resize the vector to <paramref name="newSize"/>.
        /// </summary>
        /// <param name="newSize">The new size of the vector.</param>
        public void Resize(uint newSize)
        {
            Native.Z3_ast_vector_resize(Context.nCtx, NativeObject, newSize);
        }

        /// <summary>
        /// Add the AST <paramref name="a"/> to the back of the vector. The size
        /// is increased by 1.
        /// </summary>
        /// <param name="a">An AST</param>
        public void Push(AST a)
        {
            Contract.Requires(a != null);

            Native.Z3_ast_vector_push(Context.nCtx, NativeObject, a.NativeObject);
        }

        /// <summary>
        /// Translates all ASTs in the vector to <paramref name="ctx"/>.
        /// </summary>
        /// <param name="ctx">A context</param>
        /// <returns>A new ASTVector</returns>
        public ASTVector Translate(Context ctx)
        {
            Contract.Requires(ctx != null);
            Contract.Ensures(Contract.Result<ASTVector>() != null);

            return new ASTVector(Context, Native.Z3_ast_vector_translate(Context.nCtx, NativeObject, ctx.nCtx));
        }

        /// <summary>
        /// Retrieves a string representation of the vector. 
        /// </summary>    
        public override string ToString()
        {
            return Native.Z3_ast_vector_to_string(Context.nCtx, NativeObject);
        }

        /// <summary>
        /// Translates an AST vector into an AST[]
        /// </summary>    
        public AST[] ToArray()
        {
            uint n = Size;
            AST[] res = new AST[n];
            for (uint i = 0; i < n; i++)
                res[i] = AST.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into an Expr[]
        /// </summary>    
        public Expr[] ToExprArray()
        {
            uint n = Size;
            Expr[] res = new Expr[n];
            for (uint i = 0; i < n; i++)
                res[i] = Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a BoolExpr[]
        /// </summary>    
        public BoolExpr[] ToBoolExprArray()
        {
            uint n = Size;
            BoolExpr[] res = new BoolExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (BoolExpr) Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a BitVecExpr[]
        /// </summary>    
        public BitVecExpr[] ToBitVecExprArray()
        {
            uint n = Size;
            BitVecExpr[] res = new BitVecExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (BitVecExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a ArithExpr[]
        /// </summary>    
        public ArithExpr[] ToArithExprArray()
        {
            uint n = Size;
            ArithExpr[] res = new ArithExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (ArithExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a ArrayExpr[]
        /// </summary>    
        public ArrayExpr[] ToArrayExprArray()
        {
            uint n = Size;
            ArrayExpr[] res = new ArrayExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (ArrayExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a DatatypeExpr[]
        /// </summary>    
        public DatatypeExpr[] ToDatatypeExprArray()
        {
            uint n = Size;
            DatatypeExpr[] res = new DatatypeExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (DatatypeExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a FPExpr[]
        /// </summary>    
        public FPExpr[] ToFPExprArray()
        {
            uint n = Size;
            FPExpr[] res = new FPExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (FPExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a FPRMExpr[]
        /// </summary>    
        public FPRMExpr[] ToFPRMExprArray()
        {
            uint n = Size;
            FPRMExpr[] res = new FPRMExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (FPRMExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a IntExpr[]
        /// </summary>    
        public IntExpr[] ToIntExprArray()
        {
            uint n = Size;
            IntExpr[] res = new IntExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (IntExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        /// <summary>
        /// Translates an ASTVector into a RealExpr[]
        /// </summary>    
        public RealExpr[] ToRealExprArray()
        {
            uint n = Size;
            RealExpr[] res = new RealExpr[n];
            for (uint i = 0; i < n; i++)
                res[i] = (RealExpr)Expr.Create(this.Context, this[i].NativeObject);
            return res;
        }

        #region Internal
        internal ASTVector(Context ctx, IntPtr obj) : base(ctx, obj) { Contract.Requires(ctx != null); }
        internal ASTVector(Context ctx) : base(ctx, Native.Z3_mk_ast_vector(ctx.nCtx)) { Contract.Requires(ctx != null); }

        internal class DecRefQueue : IDecRefQueue
        {
            public DecRefQueue() : base() { }
            public DecRefQueue(uint move_limit) : base(move_limit) { }
            internal override void IncRef(Context ctx, IntPtr obj)
            {
                Native.Z3_ast_vector_inc_ref(ctx.nCtx, obj);
            }

            internal override void DecRef(Context ctx, IntPtr obj)
            {
                Native.Z3_ast_vector_dec_ref(ctx.nCtx, obj);
            }
        };

        internal override void IncRef(IntPtr o)
        {
            Context.ASTVector_DRQ.IncAndClear(Context, o);
            base.IncRef(o);
        }

        internal override void DecRef(IntPtr o)
        {
            Context.ASTVector_DRQ.Add(o);
            base.DecRef(o);
        }
        #endregion
    }
}
